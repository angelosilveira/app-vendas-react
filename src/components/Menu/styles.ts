import styled, { css } from 'styled-components';

type MenuProps = {
  isOpen: boolean;
};

export const MenuButton = styled.button`
  ${({ theme }) => css`
    display: flex;

    svg {
      color: ${theme.colors.white};
      font-size: ${theme.font.sizes.xxlarge};
    }
  `}
`;

export const Submenu = styled.nav<MenuProps>`
  ${({ theme, isOpen }) => css`
    position: absolute;
    height: 100%;
    width: 320px;
    right: 0;
    z-index: 100;
    bottom: 0;
    background-color: ${theme.colors.black};
    transition: transform 0.2s ease-in;
    ${!isOpen
      ? css`
          transform: translateX(100%);
        `
      : css`
          transform: translateX(0);
        `}
  `}
`;

export const SubmenuItem = styled.div`
  ${({ theme }) => css`
    padding: 10px 20px;
    border-top: 0.5px solid rgba(255, 255, 255, 0.1);
    display: flex;
    align-items: center;

    &:first-child {
      border: 0;
    }

    svg {
      margin-right: 5px;
    }

    a,
    svg {
      color: ${theme.colors.white};
    }
  `}
`;

export const ButtonClose = styled.div`
  ${({ theme }) => css`
    text-align: right;
    padding: 0px 5px;

    svg {
      color: ${theme.colors.white};
      font-size: ${theme.font.sizes.xxlarge};
    }
  `}
`;
