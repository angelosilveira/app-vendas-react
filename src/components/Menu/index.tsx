import { useState } from 'react';
import { BiMenu, BiUser, BiGridAlt, BiX } from 'react-icons/bi';
import { Link } from 'react-router-dom';

import * as S from './styles';

const Menu = () => {
  const [open, setOpen] = useState(false);

  return (
    <>
      <S.MenuButton onClick={() => setOpen(true)}>
        <BiMenu aria-label="menu" />
      </S.MenuButton>

      <S.Submenu isOpen={open} aria-hidden={!open}>
        <S.ButtonClose>
          <button onClick={() => setOpen(false)}>
            <BiX />
          </button>
        </S.ButtonClose>

        <S.SubmenuItem>
          <BiGridAlt />
          <Link to="/reports">Dashboard</Link>
        </S.SubmenuItem>
        <S.SubmenuItem>
          <BiUser />
          <Link to="/">Clientes</Link>
        </S.SubmenuItem>
      </S.Submenu>
    </>
  );
};

export default Menu;
