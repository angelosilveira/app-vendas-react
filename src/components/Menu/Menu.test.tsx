import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from 'src/utils/tests/renderWithTheme';
import { describe, expect, test } from 'vitest';

import Menu from '.';

describe('<Menu />', () => {
  test('should render the menu', () => {
    const { container } = renderWithTheme(<Menu />);

    expect(screen.getByLabelText(/menu/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should handle open/close menu', () => {
    renderWithTheme(<Menu />);

    const menuElement = screen.getByRole('navigation', { hidden: true });
    const openMenu = screen.getByLabelText(/menu/i);

    expect(menuElement.getAttribute('aria-hidden')).toBe('true');
    expect(menuElement).toHaveStyle({ transform: 'translateX(100%)' });

    fireEvent.click(openMenu);
    expect(menuElement.getAttribute('aria-hidden')).toBe('false');
    expect(menuElement).toHaveStyle({ transform: 'translateX(0)' });
    expect(screen.getByText(/dashboard/i)).toBeInTheDocument();
    expect(screen.getByText(/clientes/i)).toBeInTheDocument();
  });
});
