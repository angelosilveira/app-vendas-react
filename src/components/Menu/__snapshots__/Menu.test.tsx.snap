// Vitest Snapshot v1

exports[`<Menu /> > should render the menu 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 svg {
  color: #FAFAFA;
  font-size: 2.8rem;
}

<button
  class="c0"
>
  <svg
    aria-label="menu"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4 6h16v2H4zm0 5h16v2H4zm0 5h16v2H4z"
    />
  </svg>
</button>
`;
